### ISSUE TITLE
Heling's research tasks & brainstorming

### ISSUE THREAD

**Issue by brando90** [HelingZhang/CoqGym#1]
here we will write tasks, brainstorming, discussions, questions, etc. for projects

Brando's availability

Monday 4-6pm (or even 7pm)
Thursdays 5-6:30pm
Fridays 4-5pm (or even 7pm)
weekends by appointment.

**Comment by brando90** [comment]
Research Task 1

- implement (or find one) MAML for pytorch (or check HW2 from here http://cs330.stanford.edu/)
- define supervised (meta) learning task for our problem. Set of proofs, geo, calc, LA then predict on unseen task e.g. mult. calc.

this will also get us to run our own ML models on Coq.

**Comment by brando90** [comment]
@HelingZhang since you asked about how to differentiate anything:

Meta-Learning with Implicit Gradients: https://arxiv.org/pdf/1909.04630
section 2.4 and lemma 1
the thing is that equation only depends on the output of Alg

**Comment by HelingZhang** [comment]
Research Task 2

Address the following initial challenges in the proof sketching scheme.
1. Explore possible ways to generate graphs with constraints, (e.g. must be a tree)
2. Explore possible representations of graphs to allow flexibility in shape and size, (e.g. adjacency matrix with a fix a max size)

**Comment by HelingZhang** [comment]
Research Task 3

1. For the base-line method: Benchmark the RL environment. Compare it with 2 other RL environments to see if it's too slow to do RL with coq. This will determine the future direction of our base-line method (RL vs supervised learning).
2. For the sketching method: Explore/Develop methods to generate tree structures conditioning on a theorem. P( y=tree | x=theorem ).

**Comment by brando90** [comment]
Meeting 4/Nov/2019:

* P(Y=Tree | X=Thm)
    * TODO: read on the the lit for phylogenetic trees
    * e-mail so some professor related this about it
* HOList:
    * https://sites.google.com/view/holist/home
* Gabrielle: RL
    * https://arxiv.org/pdf/1904.03241.pdf
    * https://arxiv.org/abs/1905.10501
    * Read/familiarize with those techniques with the goal to implement a prototype for CoqGym RL
* Heling: Supervised
    * https://arxiv.org/abs/1905.10006
    * Read/familiarize with those techniques with the goal to implement a prototype for CoqGym SL
    * future: how to do tree search with CoqHammer and our ML prover
* Brando:
    * read on the the lit for phylogenetic trees
* TODO:
    * differentiation Hammers
        * formulating a problem makes sense
        * having embeding formulas be input
    * Sketching
        * reading phylogenetic trees

**Comment by brando90** [comment]
Bruijn nameless representations

http://fsl.cs.illinois.edu/index.php/CS522_-_Programming_Language_Semantics_(Fall_2018)

**Comment by HelingZhang** [comment]
RL benchmark:

- environment         |    avg step time     |   std           |   max        | 
- Gym - Atari           |   3.54e-4               |   7.70e-5   |   4.97e-4   | 
- Gym - Robot         |   3.72e-4               |   8.07e-5   |   5.50e-4   |
- Gym - Miniworld    |    7.85e-4             |   8.71e-5   |   1.42e-3   | 
- Coq                       |    3.75e-2             |   2.43e-2   |   0.1014   | 

__________________________________________________

**Comment by brando90** [comment]
in addition, the HOList benchmark for the moment is:
```
3ms = 0.003 = 3e-3
```
but I need to do better stats on it.

**Comment by brando90** [comment]
* Meeting 22/Nov/2019
* Heling
    * Overall Goal: implement GNN for generating Coq terms
        * 1st Deliverable: 
            * GNN 0 hop for generating the encoding
                * date: 29th Friday
            * have an interface/api that is compatible with the rest of ASTactic so we can just plug this in
    * On the side, Sketching for Coq
        * Goal: reproduce adapt 1) with 2)
        * 1) Read inferring program sketches
            * https://www.overleaf.com/read/scpxvcknnttf
            * https://arxiv.org/abs/1902.06349
        * 2) Read Coq declarative language
            * https://kluedo.ub.uni-kl.de/frontdoor/deliver/index/docId/2100/file/B-065.pdf
* Brando TODO/Help:
    * find the GNN code for HOList
    * a way to automate making graphs out of Coq expressions (similar to how they do with HOL Light expressions)
        * think about algorithm for merging common nodes
        * google it
        * talk to theory friends
        * msg authors

**Comment by HelingZhang** [comment]
01/27/2020
 - train a simple encoder/decoder with reconstruction loss only
 - solve the branching problem

**Comment by brando90** [comment]
for testing the branching problem or the derivative of T I recommended this simple proof:

https://softwarefoundations.cis.upenn.edu/lf-current/Induction.html

```
Theorem plus_n_O_secondtry : ∀n:nat,
  n = n + 0.
Proof.
  intros n. destruct n as [| n'] eqn:E.
  - (* n = 0 *)
    reflexivity. (* so far so good... *)
  - (* n = S n' *)
    simpl. (* ...but here we are stuck again *)
Abort.
```

**Comment by HelingZhang** [comment]
Discuss the paper Differentiation of Blackbox Combinatorial Solvers https://arxiv.org/abs/1912.02175

**Comment by brando90** [comment]
new issue: https://github.com/brando90/DiffCoqGym/issues/1
